lisa/transformer.py

```diff
@@ -198,16 +232,25 @@ def _run_transformers(
 
 
 def run(
-    runbook_builder: RunbookBuilder, phase: str = constants.TRANSFORMER_PHASE_INIT
+    runbook_builder: RunbookBuilder,
+    phase: str = constants.TRANSFORMER_PHASE_INIT,
+    environment: Optional[Environment] = None,
 ) -> None:
     log = _get_init_logger()
 
     root_runbook_data = runbook_builder.raw_data
-    if constants.TRANSFORMER not in root_runbook_data:
+    if not root_runbook_data or constants.TRANSFORMER not in root_runbook_data:
         log.debug("no transformer found, skipped")
         return
 
     # real run
     log.debug(f"detecting or running transformers of phase '{phase}'...")
-    output_variables = _run_transformers(runbook_builder, phase=phase)
-    merge_variables(runbook_builder.variables, output_variables)
+    if environment and environment.nodes:
```

Comment:
Add code comments to explain why the environment is given some times.

Reply:
Done.